### PR TITLE
Add verify-conversions target

### LINKED ISSUE #510: Verify targets should be made in parity with CAPI
/kind bug

**What steps did you take and what happened:**
Did a comparison of verify targets used in core capi vs capg found out that couple of verify targets are missing in capg. If it makes sense, we can add those to CAPG in order to make the contributor experience more similar. This should also help improve the developer experience. 
The following section has the comparison matrix.

Missing targets is CAPG but present in CAPI:
- lint-fix
- apidiff
- verify-conversions
- verify-tiltfile
- verify-book-links
We can discuss about this, and if accepted we can add this targets too.

**Comparison matrix**
Validations executed via make

| S.no | make target | CAPI | CAPG | Function |
| --- | --- | --- | --- | --- |
| 1. |  lint-fix | ✔️ |  ❌ | runs golangci-lint with --fix |
| 2. |  apidiff | ✔️ |  ❌ | diffs API changes on PRs |
| 3. |  verify | ✔️ | ✔️ | target which runs all the other verify targets |
| 4.|  verify-modules | ✔️ | ✔️ | verifies go.mod’s are up-to-date |
| 5. |  verify-gen | ✔️ |  ✔️ | verifies all generated files are up-to-date |
| 6. |  verify-conversions | ✔️ |  ❌ | verifies all API types implement conversions |
| 7. |  verify-boilerplate | ✔️ |  ✔️ | verifies file headers |
| 8. |  verify-shellcheck | ✔️ |  ✔️ | verifies shell scripts |
| 9. |  verify-tiltfile | ✔️ |  ❌ | verifies tiltfile |
| 10. |  verify-booklinks | ✔️ |  ❌ | verifies links in books |

**Environment:**

- Cluster-api version:
- Minikube/KIND version:
- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by sayantani11**
**What type of PR is this?**

**What this PR does / why we need it**:
verify-conversions verifies all API types implement conversions

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #510 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
verify-conversions target added
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -251,6 +252,10 @@ $(RELEASE_NOTES): ## Build release notes.
 $(GO_APIDIFF): ## Build go-apidiff from tools folder.
 	GOBIN=$(TOOLS_BIN_DIR) $(GO_INSTALL) github.com/joelanford/go-apidiff $(GO_APIDIFF_BIN) $(GO_APIDIFF_VER)
 
+conversion-verifier: $(CONVERSION_VERIFIER) go.mod go.sum ## fetch CAPI's conversion verifier
+$(CONVERSION_VERIFIER): go.mod
```
Comment (sbueringer) [inline Makefile:256]:
```suggestion
$(CONVERSION_VERIFIER): $(TOOLS_DIR)/go.mod ## fetch CAPI's conversion verifier
```
I think we don't need / shouldn't add the conversion-verifier target

File: Makefile [hunk 2]
```diff
@@ -289,7 +294,7 @@ lint-full: $(GOLANGCI_LINT) ## Run slower linters to detect possible issues
 
 .PHONY: modules
 modules: ## Runs go mod to ensure proper vendoring.
-	go mod tidy
+	cd $(TOOLS_DIR); go mod tidy
```
Comment (sbueringer) [inline Makefile:297]:
```suggestion
	go mod tidy
	cd $(TOOLS_DIR); go mod tidy
```
we need both

File: hack/tools/go.mod [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+	k8s.io/apiextensions-apiserver v0.23.0 // indirect
+	k8s.io/apimachinery v0.23.0 // indirect
+	k8s.io/klog/v2 v2.30.0 // indirect
+	k8s.io/utils v0.0.0-20210930125809-cb0fa318a74b // indirect
+	sigs.k8s.io/controller-tools v0.8.0 // indirect
+	sigs.k8s.io/json v0.0.0-20211020170558-c049b76a60c6 // indirect
+	sigs.k8s.io/structured-merge-diff/v4 v4.2.0 // indirect
+        sigs.k8s.io/cluster-api v1.1.3
+        sigs.k8s.io/cluster-api/test v1.1.3
+)
```
Comment (sbueringer) [inline hack/tools/go.mod:36]: Please add this to the end of the go.mod:
```
replace (
	sigs.k8s.io/cluster-api => sigs.k8s.io/cluster-api v1.1.3
	sigs.k8s.io/cluster-api/test => sigs.k8s.io/cluster-api/test v1.1.3
)
```

then run `make modules` (after the makefile fixes have been applied)

**Comment by sbueringer** [comment]
These fixes should fix the current issue of the verify job. Afterwards we should get some findings from the conversion-verifier. But let's talk about them after the first fixes

**Comment by sbueringer** [comment]
Hm, not sure what is going on. The verify target should fail now but it doesn't. It fails as expected when I run it on my machine.

Can you add a `cd $(ROOT_DIR)` at the beginning of the verify-conversion target? Maybe it's run in the wrong directory somehow

**Comment by sayantani11** [comment]
> Hm, not sure what is going on. The verify target should fail now but it doesn't. It fails as expected when I run it on my machine.
> 
> Can you add a `cd $(ROOT_DIR)` at the beginning of the verify-conversion target? Maybe it's run in the wrong directory somehow

ok

**Comment by sayantani11** [comment]
> Hm, not sure what is going on. The verify target should fail now but it doesn't. It fails as expected when I run it on my machine.
> 
> Can you add a `cd $(ROOT_DIR)` at the beginning of the verify-conversion target? Maybe it's run in the wrong directory somehow

Looking into it

**Comment by sayantani11** [comment]
@sbueringer Yep!! Guess that was the issue

**Comment by sayantani11** [comment]
> @sayantani11 Found the issue.
> 
> The functions which make verify-conversion succeed are already implemented on the main branch. ([PR](https://github.com/kubernetes-sigs/cluster-api-provider-gcp/pull/559/files))
> 
> I was locally testing this PR, while Prow is testing the result of merging this PR into main.
> 
> Can you please squash the commits and rebase onto main?

Yes sure!!!

**Review by cpanato — APPROVED** [review]
Thanks!
/approve

**Pull request merged**